text/3851-supertrait-auto-impls.md

```diff
@@ -0,0 +1,975 @@
+// then it is impossible for the `impl BaseFunction` to compile.
+
+impl<T> ManagementExtension for MangementInterface<T>
+where T: ManagementHandle
+{
+    extern impl BaseFunction;
+    fn management_interface(&self) { .. }
+}
+```
+
+## Why explicit opt-in/out for marker traits and traits with only default items?
+
+Marker supertraits and supertraits with `default` items can be easily overlooked when users write subtrait implementations. They would register too little signal for the reader to recognise the significance of traits of these kinds. For this reason, we bias towards asking users to provide clear syntatical signals through `auto impl MarkerTrait` or `extern impl MarkerTrait`, so that the automatic derivation of such traits is easily recognisable and provides obvious site for documentation in case justification is waranted.
+
+```rust
+// It is almost always a good idea to explain
+// why a trait like below is implemented on a type.
+trait MarkerTrait {}
+
+trait Supertrait: MarkerTrait {
+    auto impl MarkerTrait;
+}
+
+impl Supertrait for MyType {
+    // Here it is a good place to explain
+    // why MyType: MarkerTrait
+    auto impl MarkerTrait;
+    // or otherwise `extern impl MarkerTrait;` is required
+}
+```
+
+### What about checking whether another impl exists to decide automatic `impl`-filling?
+
+We could potentially determine whether the opt-out is used based on whether an `impl` of the supertrait exists, but we prefer not to. We have existing mechanism to determine the specialisation and implementation overlaps. Whether a supertrait `impl` overlaps or not, is not the concern of this proposal.
+
+If we would deduce whether an `auto impl` should be effected, there could present a hazard that silently changes the program behaviour.
+
+```rs
+trait MyTrait { default fn .. }
+
+trait AutoMyTrait: MyTrait {
+    auto impl MyTrait;
+}
+
+trait MyOtherTrait { .. }
+
+struct Foo;
+
+// Suppose we allow users to elide the `auto impl`/`extern impl` directive
+// and we deduce based on some applicability criterion ...
+impl AutoMyTrait for Foo {} // <-- generates MyTrait
+
+// Suppose this item is added at one point of time
+impl<T: MyOtherTrait> MyTrait for T { .. }
+
+impl MyOtherTrait for Foo {}
+
+// QUESTION: which `impl MyTrait for Foo` fulfills the bound `Foo: MyTrait`?
+// Should it be the `auto impl MyTrait;` with default items?
+// Should it be the `impl<T: MyOtherTrait> MyTrait for T` with `T := Foo`?
+```
+
+## Why is a hard error on ambiguity acceptable to us?
+
+We hold the basic assumption that most associated items of a trait have sensible names. We would rather advise that one shall avoid name clashes and ambiguity through better, future-oriented trait designs.
+
+In any case, `auto impl Trait { .. }` blocks still remains available for cases where ambiguity is unavoidable or favorable in niche scenario.
+
+## Why this naming?
+
+It is still up for discussion.
+
+## Why use `unsafe` in this way
```

Comment:
The RFC no longer uses `unsafe` in a surprising way, so I think this section can be dropped.

Reply:
I think the RFC still uses `unsafe` in a surprising way: see [this thread](https://github.com/rust-lang/rfcs/pull/3851#discussion_r2376681991)